### PR TITLE
Bug fixes for 1.0.11

### PR DESCRIPTION
**Pull request by tushargupta51**
Multiple bug fixes:
1. #282: initialize anonymousEndpoints with a default value
2. #278: replace login() by loginHandler() so that start page is saved
3. #232 and #130: restore the start page url correctly, so that we do not loose any query string parameters
4. #201: update location hash using $location.hash, and use event.preventDefault when renewing token so browser refresh does not happen.

### CODE REVIEW

**Comment by tushargupta51** [comment]
@polita @weijjia

File: lib/adal-angular.js [hunk 1]
```diff
@@ -220,16 +212,7 @@
...
                         if (isADLoginRequired(toState, _adal.config)) {
-                            if (!_oauthData.isAuthenticated && !_adal._renewActive) {
-                                // $location.$$url is set as the page we are coming from
-                                // Update it so we can store the actual location we want to
-                                // redirect to upon returning
-                                $location.$$url = toState.url;
-
-                                // Parameters are not stored in the url on stateChange so
-                                // we store them
-                                _adal._saveItem(_adal.CONSTANTS.STORAGE.START_PAGE_PARAMS, JSON.stringify(toParams));
```
Comment (polita) [inline lib/adal-angular.js]: Have you removed the constant now that we don't use it?

File: lib/adal-angular.js [hunk 2]
```diff
@@ -125,16 +125,7 @@
...
-                                            if (paramsJSON) {
-                                                // If params were stored redirect to the page and then 
-                                                // initialize the params
-                                                var loginStartPageParams = JSON.parse(paramsJSON);
-                                                $location.url(loginStartPage).search(loginStartPageParams);
-                                            } else {
-                                                $location.url(loginStartPage);
-                                            }
+                                            $location.url(loginStartPage);
                                         }
```
Comment (polita) [inline lib/adal-angular.js]: We chatted about this, but we should make sure we're saving off and adding back on any hash string.

Comment (tushargupta51) [inline lib/adal-angular.js]: Yes, it preserves the hash component too.

File: lib/adal-angular.js [hunk 3]
```diff
@@ -328,33 +311,40 @@
...
-                        } else {
+                        }
+                        else {
                             // Cancel request if login is starting
                             if (authService.loginInProgress()) {
-                                authService.info('login already started.');
-                                return $q.reject('login in progress, cancelling the request');
-                            } else {
+                                authService.info('login is in progress.');
+                                config.data = 'login in progress, cancelling the request';
```
Comment (polita) [inline lib/adal-angular.js]: Nit: we should be consistent in our messages: either capitalize the first word for all of them, or have them all lowercase. Below, "Error" is capitalized.

File: lib/adal.js [hunk 4]
```diff
@@ -662,7 +662,7 @@ var AuthenticationContext = (function () {
 
             if (requestInfo.requestType === this.REQUEST_TYPE.LOGIN) {
                 this._loginInProgress = false;
-                this._saveItem(this.CONSTANTS.STORAGE.LOGIN_ERROR, requestInfo.parameters.errorDescription);
+                this._saveItem(this.CONSTANTS.STORAGE.LOGIN_ERROR, requestInfo.parameters.error_description);
```
Comment (polita) [inline lib/adal.js]: Why did we add an underscore between "error" and "description"? It looks like the predominant style is to use camel casing.

Comment (tushargupta51) [inline lib/adal.js]: It should match with what AAD returns in the url. AAD uses error_description.

Comment (polita) [inline lib/adal.js]: Got it, thanks.

File: lib/adal-angular.js [hunk 5]
```diff
@@ -83,11 +83,8 @@
                         var requestInfo = _adal.getRequestInfo(hash);
                         _adal.saveTokenFromHash(requestInfo);
 
-                        if ($location.$$html5) {
-                            $window.location = $window.location.origin + $window.location.pathname;
-                        } else {
-                            $window.location.hash = '';
-                        }
+                        // resetting hash to remove fragments sent by AAD
+                        $location.hash('');
```
Comment (weijjia) [inline lib/adal-angular.js]: i remembered that developer were complaining that the hash change will trigger that location event, will this also cause the problem?

Comment (tushargupta51) [inline lib/adal-angular.js]: @weijjia $location.hash('') will not trigger another location change. Updating $window.location was doing that.

**Comment by weijjia** [comment]
:shipit:

**Pull request merged**